### PR TITLE
New Style Button Command and New Alpha Button Command

### PR DESCRIPTION
**Pull request by DoctorRyan**
This is an update to the palette viewer that adds a more convenient plus button for adding colors and a more clear Color 0 button to avoid people confusing this with white color.  The original author was Turtletooth.  Thanks to him for writing the code.

### CODE REVIEW

**Comment by RodneyBaker** [comment]
@shun-iwasawa  As there are several changes to the palette window included in this PR outside of adding the new style button and changing color_0 visual appearance it is appropriate to get your feedback on how this might effect current users, especially those in Japanese studios.

**Comment by shun-iwasawa** [comment]
@DoctorRyan 
> The original author was Turtletooth. 

First of all, please let me clarify, did you get a permission from him to use his code or is it released under certain license?

**Comment by RodneyBaker** [comment]
Shun,
@turtletooth (Jeremy Bullock) has stated this code is under the same BSD license as Opentoonz.

**Comment by turtletooth** [comment]
I don't mind if anything from Tahoma gets back-ported to OpenToonz.

**Comment by turtletooth** [comment]
However, I would also bring in my change to the colors of the checkerboard to make the colors more in line with other software.
colors1

File: toonz/sources/include/toonzqt/paletteviewergui.h [hunk 1]
```diff
@@ -116,6 +116,9 @@ class DVAPI PageViewer final : public QFrame, public TSelection::View {
   NameDisplayMode getNameDisplayMode() const { return m_nameDisplayMode; }
   void setNameDisplayMode(NameDisplayMode mode);
 
+  void toggleShowStyleIndex();
```
Comment (turtletooth) [inline toonz/sources/include/toonzqt/paletteviewergui.h:119]: I'm leaving comments for Ryan:
This makes the style indexes in the corner toggle-able.  I'm not sure that is desired for OpenToonz.

File: toonz/sources/toonzqt/paletteviewer.cpp [hunk 2]
```diff
@@ -105,7 +105,7 @@ PaletteViewer::PaletteViewer(QWidget *parent, PaletteViewType viewType,
   // Create pageView
   m_pageViewerScrollArea = new QScrollArea();
   m_pageViewerScrollArea->setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
-  m_pageViewerScrollArea->setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOn);
+  m_pageViewerScrollArea->setVerticalScrollBarPolicy(Qt::ScrollBarAsNeeded);
```
Comment (turtletooth) [inline toonz/sources/toonzqt/paletteviewer.cpp:108]: Most of the changes in this file are beyond the scope of this PR.

File: toonz/sources/toonzqt/paletteviewergui.cpp [hunk 3]
```diff
@@ -41,6 +41,8 @@
 TEnv::IntVar EnvSoftwareCurrentFontSize_StyleName(
     "SoftwareCurrentFontSize_StyleName", 11);
 
+TEnv::IntVar ShowStyleIndex("ShowStyleIndex", 0);
```
Comment (turtletooth) [inline toonz/sources/toonzqt/paletteviewergui.cpp:44]: Unnecessary if the indexes are going to be permanent.

File: toonz/sources/toonzqt/paletteviewergui.cpp [hunk 4]
```diff
@@ -138,16 +140,16 @@ PageViewer::PageViewer(QWidget *parent, PaletteViewType viewType,
   ViewMode defaultChipSize;
   switch (m_viewType) {
   case LEVEL_PALETTE:
-    defaultChipSize = LargeChips;
+    defaultChipSize = SmallChips;
```
Comment (turtletooth) [inline toonz/sources/toonzqt/paletteviewergui.cpp:143]: I'd ask before changing the default chip sizes.

File: toonz/sources/toonzqt/paletteviewergui.cpp [hunk 5]
```diff
@@ -226,6 +228,20 @@ int PageViewer::getCurrentStyleIndex() const {
   return getPaletteHandle()->getStyleIndex();
 }
 
+//-----------------------------------------------------------------------------
+
+void PageViewer::toggleShowStyleIndex() {
```
Comment (turtletooth) [inline toonz/sources/toonzqt/paletteviewergui.cpp:233]: Again unnecessary if keeping the style indexes.

File: toonz/sources/toonzqt/paletteviewergui.cpp [hunk 6]
```diff
@@ -715,10 +750,12 @@ void PageViewer::paintEvent(QPaintEvent *e) {
         if ((int)style->getMainColor().m !=
             (int)style->getMainColor().maxChannelValue) {
           QRect bottomRect = chipRect;
-          if (m_viewMode == LargeChips) {
-            bottomRect.adjust(0, bottomRect.height() - 12, 0, 0);
-          } else
-            bottomRect.adjust(0, bottomRect.height() - 6, 0, 0);
+          if (styleIndex != 0) {
```
Comment (turtletooth) [inline toonz/sources/toonzqt/paletteviewergui.cpp:753]: I'd remove this change.

File: toonz/sources/toonzqt/paletteviewergui.cpp [hunk 7]
```diff
@@ -766,10 +803,7 @@ void PageViewer::paintEvent(QPaintEvent *e) {
       }
 
       // draw style name
-      QFont preFont = p.font();
-      QFont tmpFont = p.font();
-
-      if (m_viewMode != SmallChips) {
+      if (m_viewMode != SmallChips && styleIndex != 0) {
```
Comment (turtletooth) [inline toonz/sources/toonzqt/paletteviewergui.cpp:806]: Remove this change too.

File: toonz/sources/toonzqt/paletteviewergui.cpp [hunk 8]
```diff
@@ -825,13 +858,15 @@ void PageViewer::paintEvent(QPaintEvent *e) {
...
       tmpFont.setItalic(false);
       p.setFont(tmpFont);
-      int indexWidth = fontMetrics().width(QString().setNum(styleIndex)) + 4;
-      QRect indexRect(chipRect.bottomRight() + QPoint(-indexWidth, -14),
-                      chipRect.bottomRight());
-      p.setPen(Qt::black);
-      p.setBrush(Qt::white);
-      p.drawRect(indexRect);
-      p.drawText(indexRect, Qt::AlignCenter, QString().setNum(styleIndex));
+      if (ShowStyleIndex == 1) {
```
Comment (turtletooth) [inline toonz/sources/toonzqt/paletteviewergui.cpp:861]: This can be removed too.

File: toonz/sources/toonzqt/paletteviewergui.cpp [hunk 9]
```diff
@@ -653,6 +671,23 @@ void PageViewer::paintEvent(QPaintEvent *e) {
       // toggle link
       drawToggleLink(p, chipRect, m_page->getStyle(i));
     }
+    if (!m_page->getPalette()->isLocked()) {
```
Comment (turtletooth) [inline toonz/sources/toonzqt/paletteviewergui.cpp:674]: This is where the new style button is drawn in list view.  Keep this.

File: toonz/sources/toonzqt/paletteviewergui.cpp [hunk 10]
```diff
@@ -880,6 +915,24 @@ void PageViewer::paintEvent(QPaintEvent *e) {
       // draw link indicator
       drawToggleLink(p, chipRect, style);
     }
+    // draw new style chip
```
Comment (turtletooth) [inline toonz/sources/toonzqt/paletteviewergui.cpp:918]: This is where the new style button is drawn.  Keep this.

File: toonz/sources/toonzqt/paletteviewergui.cpp [hunk 11]
```diff
@@ -955,17 +1008,20 @@ void PageViewer::mousePressEvent(QMouseEvent *event) {
   }
   m_dragStartPosition = pos;
   if (indexInPage < 0 || indexInPage >= getChipCount()) {
-    // l'utente ha fatto click fuori dai color chip. vuole deselezionare tutto
-    // (lasciando la selezione attiva, per un eventuale paste)
-    m_styleSelection->select(pageIndex);
-    m_styleSelection->makeCurrent();
-
+    if (indexInPage == getChipCount() && !m_page->getPalette()->isLocked()) {
```
Comment (turtletooth) [inline toonz/sources/toonzqt/paletteviewergui.cpp:1011]: This is the logic for pressing the new button.  Keep this.

File: toonz/sources/toonzqt/paletteviewergui.cpp [hunk 12]
```diff
@@ -1325,15 +1382,24 @@ bool PageViewer::event(QEvent *e) {
...
       if (style) {
         int styleIndex = m_page->getStyleId(indexInPage);
-        toolTip        = "#" + QString::number(styleIndex) + " " +
-                  QString::fromStdWString(style->getName());
-
-        int shortcutKey = m_page->getPalette()->getStyleShortcut(styleIndex);
-        if (shortcutKey > 0)
-          toolTip += QString::fromStdWString(std::wstring(L" (") +
-                                             (wchar_t)shortcutKey + L")");
+        if (styleIndex == 0) {
```
Comment (turtletooth) [inline toonz/sources/toonzqt/paletteviewergui.cpp:1385]: I'd remove these changes.

File: toonz/sources/toonzqt/paletteviewergui.cpp [hunk 13]
```diff
@@ -1325,15 +1382,24 @@ bool PageViewer::event(QEvent *e) {
...
+                    QString::fromStdWString(style->getName());
+
+          int shortcutKey = m_page->getPalette()->getStyleShortcut(styleIndex);
+          if (shortcutKey > 0)
+            toolTip += QString::fromStdWString(std::wstring(L" (") +
+                                               (wchar_t)shortcutKey + L")");
+        }
       }
     }
+    if (indexInPage == m_page->getStyleCount()) {
```
Comment (turtletooth) [inline toonz/sources/toonzqt/paletteviewergui.cpp:1400]: Keep this.

**Comment by DoctorRyan** [comment]
Thanks for the review Turtle.  I apologize for posting so late.  Mr. @shun-iwasawa   As they have stated I was given his blessing to try to import over some of his changes to Opentoonz.  Still some work to be done that I will be sure to get done.  My goal is to try and help make OT more user friendly if possible.  Any suggestions or feedback is most appreciated.  A pleasure to finally meet you.

**Comment by manongjohn** [comment]
Build errors that need to be addressed:

C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(133): error C2065: 'm_newPageToolbar': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(134): error C2065: 'm_newPageToolbar': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(134): error C2227: left of '->clear' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(134): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(135): error C2065: 'm_newPageToolbar': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(135): error C2227: left of '->setMovable' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(135): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(136): error C2065: 'm_newPageToolbar': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(136): error C2227: left of '->setIconSize' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(136): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(137): error C2065: 'm_newPageToolbar': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(137): error C2227: left of '->setLayoutDirection' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(137): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(142): error C2065: 'm_newPageToolbar': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(145): error C2065: 'm_newPageToolbar': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(145): error C2227: left of '->addAction' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(145): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(151): error C2065: 'm_newPageToolbar': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(295): error C3861: 'updatePaletteMenu': identifier not found [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(442): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(443): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(443): error C2227: left of '->setLayoutDirection' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(443): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(445): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(451): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(456): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(456): error C2227: left of '->addAction' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(456): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(464): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(464): error C2227: left of '->addSeparator' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(464): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(476): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(476): error C2227: left of '->addAction' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(476): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(489): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(489): error C2227: left of '->addSeparator' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(489): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(490): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(490): error C2227: left of '->addAction' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(490): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(493): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(493): error C2227: left of '->addSeparator' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(493): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(499): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(499): error C2227: left of '->addAction' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(499): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(505): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(505): error C2227: left of '->addAction' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(505): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(508): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(513): error C3861: 'updatePaletteMenu': identifier not found [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(568): error C2065: 'm_newPageToolbar': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(568): error C2227: left of '->actions' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(568): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(605): error C2039: 'updatePaletteMenu': is not a member of 'PaletteViewer' [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\include\toonzqt/studiopaletteviewer.h(30): note: see declaration of 'PaletteViewer'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(606): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(608): error C2065: 'm_viewMode': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(608): error C2227: left of '->actions' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(608): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(609): error C3861: 'getPalette': identifier not found [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(610): error C2065: 'm_pageViewer': undeclared identifier [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(610): error C2227: left of '->getViewMode' must point to class/struct/union/generic type [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
  C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(610): note: type is 'unknown-type'
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(614): error C3861: 'tr': identifier not found [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(615): error C3861: 'tr': identifier not found [C:\projects\opentoonz\toonz\x64\toonzqt\toonzqt.vcxproj]
C:\projects\opentoonz\toonz\sources\toonzqt\paletteviewer.cpp(616): error C3861: 'tr': identifier not found

**Comment by turtletooth** [comment]
@shun-iwasawa I updated the license file at Tahoma:
https://github.com/turtletooth/tahoma/blob/master/LICENSE.txt

**Comment by DoctorRyan** [comment]
Sorry, my commit history looks like a train wreck.  However, after breaking things 10345 times I am back on track with the commit.  Removed all unnecessary lines of code, fixed changes to UI that didn't need to be changed as well.  Tested on my local and looks good so far.  I will take care of the checkboard pattern to make it not so garish next.

**Comment by DoctorRyan** [comment]
@manongjohn Did you take a look and see if that cleared up the issues for you?

**Pull request closed without merging**